## Modeling and Conclusion

Having processed our data for machine learning modeling, we will now train and evaluate several models which would predict the housing sales price in Ames Iowa. 

### Import the libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math

from sklearn.linear_model import (LinearRegression, 
                                  LassoCV, 
                                  RidgeCV, 
                                  ElasticNetCV, 
                                  Lasso)

from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

In [2]:
pd.set_option("display.max_rows", 100)
pd.set_option("display.max_columns", 100)

### Import the cleaned and processed train and test dataset

In [3]:
df_train = pd.read_csv('../data/train_processed.csv')
df_test = pd.read_csv('../data/test_processed.csv')

### Split the train data into train and validation sets

In [4]:
# Drop 'saleprice' from both the test and train data sets
X = df_train.drop(columns=['id', 'pid', 'saleprice'])
y = df_train['saleprice']

In [5]:
#Number of features

X.shape[1]

72

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
                                    X,
                                    y,
                                    test_size=0.3,
                                    random_state= 89
)

### Establish the baseline score

To establish the baseline RMSE, we assume that the predicted target value `y_base` is the mean sale price

In [7]:
# Create an array with all predictions being the mean sale price
sales_mean = np.repeat(np.mean(df_train['saleprice']), df_train.shape[0])
y_base = pd.Series(sales_mean)

In [8]:
# Calculate the baseline RMSE
round((mean_squared_error(y, y_base))**0.5, 2)

78632.44

### Standardize X_train and X_test

We will standardize the continuous variables on X_train and X_test

In [9]:
# List of continuous variables for scaling

con_var = ['1st_flr_sf', 'bsmt_exposure', 'bsmt_qual', 'bsmtfin_sf_1', 
           'bsmtfin_type_1', 'exter_qual', 'fence', 'fireplace_qu', 'full_bath',
           'garage_area', 'garage_cars', 'garage_finish', 'gr_liv_area', 'heating_qc',
           'kitchen_qual', 'lot_area', 'lot_frontage_reg_imputed', 'lot_shape', 'mas_vnr_area',
           'neigh_group', 'open_porch_sf', 'overall_qual', 'total_bsmt_sf', 'totrms_abvgrd', 
           'wood_deck_sf', 'year_built', 'year_remod_add']

In [10]:
# Isolate continuous variables for scaling
X_train_cont = X_train[[var for var in con_var if var in df_train.columns]]
X_test_cont = X_test[[var for var in con_var if var in df_test.columns]]

In [11]:
# Instantiate the StandardScaler
ss = StandardScaler()

X_train_ss_cont = ss.fit_transform(X_train_cont)
X_test_ss_cont = ss.transform(X_test_cont)

In [12]:
# Drop the orginal continuous variables from X_train and join back the scaled ones
X_train_ss = X_train.drop(
                 columns=[var for var in con_var if var in df_train.columns]
             ).reset_index(drop=True
                          ).join(pd.DataFrame(X_train_ss_cont, 
                                              columns=[var for var in con_var if var in df_train.columns]))

In [13]:
# Drop the orginal continuous variables from X_test and join back the scaled ones
X_test_ss = X_test.drop(
                 columns=[var for var in con_var if var in df_train.columns]
             ).reset_index(drop=True
                          ).join(pd.DataFrame(X_test_ss_cont, columns=[var for var in con_var if var in df_train.columns]))

### Perform Linear Regression

The first model we will train is Linear Regression

In [14]:
# Instantiate and fit the Linear Regression
lr = LinearRegression()
lr.fit(X_train_ss, y_train)

LinearRegression()

We will evaluate the cross-validated RMSE score against that of the holdout set

In [15]:
#Cross Validated RMSE of training set
round(((-cross_val_score(
    lr,
    X_train_ss,
    y_train,
    cv=10,
    scoring='neg_mean_squared_error'
).mean())**0.5), 2)

25941.91

In [16]:
round((mean_squared_error(y_test, lr.predict(X_test_ss)))**0.5, 2)

28752.18

The RMSE on the training set is lower than that of the holdout set, which is within expectation

Two thresholds were used to tune and evaluate the performance of each models in the previous notebook:

1. Cardinal Threshold: used to control the number categories in nominal variables. Categories are sorted by value counts and normalized. Any categories that are beyond the threshold are deemed to be too small and will be grouped together as a single category called 'Others'. This helps to reduce the number of features for modeling

2. Correlation Threshold: Highly correlated pair of variables beyond the stated threshold are filtered out and made to drop one of them. This reduces the multicollinearity problem and also the number of features

The following is the evaluated perfomance of each model. The best model is model 4 based on Holdout RMSE score of 28717.78

| Models | Thresholds                     | Features | CV RMSE | Holdout RMSE |
|--------|--------------------------------|----------|---------|--------------|
| 1      | cardinal:0.85 correlation:0.8   | 71       | 28739.24   | 31313.84        |
| 2      | cardinal:0.9 correlation:0.8  | 77       | 28695.65   | 31128.54        |
| 3      | cardinal:0.8 correlation:0.8   | 68       | 28966.67   | 31264.25        |
| 4      | cardinal:0.8 correlation:0.9 | 72       | 25928.04   | 28717.78        |

In [17]:
lr_coef = pd.DataFrame(lr.coef_, index = X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

lr_coef

Coefficient
bsmt_qual                 1.237077e+17
1st_flr_sf                1.237077e+17
bsmtfin_type_1            1.237077e+17
exter_qual                1.237077e+17
bsmt_exposure             1.237077e+17
exterior_2nd_VinylSd      5.835209e+16
exterior_2nd_PreCast      5.835209e+16
lot_config_Others         1.570802e+16
exterior_1st_MetalSd      1.278228e+16
exterior_2nd_HdBoard      1.278228e+16
exterior_1st_Wd Sdng      1.278228e+16
exterior_1st_PreCast      1.278228e+16
exterior_1st_HdBoard      1.278228e+16
fireplace_qu              1.041853e+16
fireplaces                1.041853e+16
exterior_2nd_Others       8.896957e+15
exterior_2nd_Other        8.896957e+15
fence                     6.763366e+15
exterior_2nd_Wd Sdng      6.763366e+15
house_style_1Story        6.625810e+15
gr_liv_area               6.118981e+15
kitchen_qual              6.118981e+15
garage_type_Others        6.118981e+15
house_style_Others        6.118981e+15
mas_vnr_area              6.118981e+15
garage_area               6.118981e+15
lot_frontage_reg_imputed  4.961153e+15
garage_finish             2.416528e+15
exterior_1st_Others       1.916117e+15
lot_area                  1.851798e+15
heating_qc                1.415391e+15
ms_subclass_85            1.927600e+04
roof_style_Others         1.838400e+04
ms_subclass_20            1.118400e+04
open_porch_sf             9.874000e+03
ms_subclass_40            7.524000e+03
ms_subclass_Others        7.024000e+03
total_bsmt_sf             6.696000e+03
ms_zoning_Others          6.144000e+03
overall_qual              5.876000e+03
ms_subclass_50            5.040000e+03
ms_subclass_180           4.464000e+03
ms_subclass_70            4.374000e+03
year_remod_add            3.600000e+03
year_built                3.136000e+03
ms_subclass_160           2.609500e+03
totrms_abvgrd             1.993000e+03
ms_subclass_80            1.378000e+03
wood_deck_sf              1.160000e+03
ms_subclass_90            1.000000e+03
ms_zoning_RL              6.200000e+02
ms_subclass_190           5.200000e+02
roof_style_Gable          3.480000e+02
ms_subclass_45            3.440000e+02
neigh_group              -3.580000e+02
ms_subclass_30           -3.880000e+02
ms_subclass_75           -6.420000e+02
ms_subclass_60           -2.316000e+03
exterior_2nd_MetalSd     -3.120000e+03
foundation_Others        -4.370789e+15
garage_type_Attchd       -6.801990e+15
mas_vnr_type_None        -8.134321e+15
mas_vnr_type_CBlock      -8.134321e+15
lot_config_Inside        -8.156071e+15
exterior_1st_VinylSd     -8.951853e+15
lot_shape                -1.324816e+16
ms_subclass_120          -1.687908e+16
mas_vnr_type_Others      -1.687908e+16
foundation_PConc         -2.064120e+16
full_bath                -2.064120e+16
bsmtfin_sf_1             -2.326717e+16
garage_cars              -2.721459e+16

In [18]:
#Number of coefficient remaining after fitting
lr_coef[lr_coef!=0].notna().sum()

Coefficient    72
dtype: int64

However, we are seeing wildly overblown coefficients on the Linear Regression. There are also  minor reduction in the number of features before and after fitting the training set(from 75 to 71). We will need to use Regression with regularization to penalize these coefficients.

### Perform Ridge Regression

We will now train a Ridge Regression based on the best Cardinality Threshold and Correlation Threshold obtained for Linear Regression 

In [19]:
ridge = RidgeCV(alphas=np.logspace(0, 5, 100))
ridge.fit(X_train_ss, y_train)

round(ridge.alpha_, 2)

29.15

In [20]:
#RMSE of training set
round((-cross_val_score(
    ridge,
    X_test_ss,
    y_test,
    cv=10,
    scoring='neg_mean_squared_error'
).mean())**0.5, 2)

29336.05

In [21]:
round((mean_squared_error(y_test, ridge.predict(X_test_ss)))**0.5, 2)

28669.68

In [22]:
ridge_coef = pd.DataFrame(ridge.coef_, 
                          index = X.columns, 
                          columns=['Coefficient']
                         ).sort_values(by='Coefficient', 
                                       ascending=False)
ridge_coef

Coefficient
roof_style_Others         17458.215965
ms_subclass_85            17057.979821
ms_subclass_20            10703.064477
open_porch_sf              9603.545389
ms_subclass_40             7764.085021
ms_subclass_Others         7045.903158
total_bsmt_sf              6529.132510
ms_zoning_Others           6112.364167
overall_qual               5748.414206
foundation_PConc           5563.234079
gr_liv_area                4507.821107
ms_subclass_70             4462.141647
ms_subclass_50             4416.644156
ms_subclass_180            4221.677308
ms_subclass_160            3834.801949
exterior_2nd_HdBoard       3785.387170
bsmt_qual                  3607.046613
kitchen_qual               3389.294157
year_remod_add             3285.190694
ms_subclass_120            2897.420895
totrms_abvgrd              2616.904943
exterior_2nd_VinylSd       2578.282858
year_built                 2488.180701
exterior_2nd_Others        2379.001769
garage_type_Others         1886.377241
exterior_1st_MetalSd       1879.733595
ms_subclass_80             1329.530334
wood_deck_sf               1225.843964
house_style_Others         1084.573444
ms_subclass_90             1041.681738
ms_subclass_190             986.230186
fence                       980.332015
ms_zoning_RL                824.197679
1st_flr_sf                  782.421857
bsmt_exposure               531.535599
roof_style_Gable            464.959751
ms_subclass_45              309.672046
mas_vnr_type_None           208.183742
fireplace_qu                172.486298
house_style_1Story            0.000000
garage_cars                   0.000000
bsmtfin_sf_1                  0.000000
exterior_1st_Others           0.000000
exterior_1st_VinylSd          0.000000
lot_area                      0.000000
foundation_Others             0.000000
lot_config_Inside             0.000000
garage_finish                 0.000000
lot_config_Others             0.000000
garage_type_Attchd            0.000000
heating_qc                    0.000000
lot_frontage_reg_imputed      0.000000
lot_shape                     0.000000
fireplaces                 -172.486298
ms_subclass_30             -176.828685
mas_vnr_type_CBlock        -208.183742
ms_subclass_75             -415.266786
neigh_group                -481.890283
exterior_1st_Wd Sdng       -744.351949
exterior_2nd_Wd Sdng       -980.332015
exterior_2nd_MetalSd      -1572.226259
exterior_1st_PreCast      -1841.613889
ms_subclass_60            -1947.056691
bsmtfin_type_1            -2158.349341
exterior_2nd_Other        -2379.001769
exterior_2nd_PreCast      -2578.282858
exter_qual                -2762.654728
mas_vnr_type_Others       -2897.420895
exterior_1st_HdBoard      -3079.154927
mas_vnr_area              -4671.899904
full_bath                 -5563.234079
garage_area               -6196.166046

In [23]:
#Number of coefficient remaining after fitting
ridge_coef[ridge_coef!=0].notna().sum()

Coefficient    58
dtype: int64

We see that the lasso regularization has reduced the previously overblown coefficients. The number of coefficients also reduced from 75 to 58

### Perform Lasso Regression

We will now train a Lasso Regression based on the best Cardinality Threshold and Correlation Threshold obtained for Linear Regression 

In [24]:
lasso = LassoCV(n_alphas=1000)
lasso.fit(X_train_ss, y_train)

round(lasso.alpha_, 2)

92.43

In [25]:
round((-cross_val_score(
    lasso,
    X_train_ss,
    y_train,
    cv=10,
    scoring='neg_mean_squared_error'
).mean())**0.5, 2)

25831.85

In [26]:
round((mean_squared_error(y_test, lasso.predict(X_test_ss)))**0.5, 2)

28636.59

In [27]:
lasso_coef = pd.DataFrame(lasso.coef_, 
                          index = X.columns, 
                          columns=['Coefficient']
                         ).sort_values(by='Coefficient', ascending=False)

lasso_coef

Coefficient
ms_subclass_85            1.952738e+04
roof_style_Others         1.824289e+04
foundation_PConc          1.144170e+04
ms_subclass_20            1.085921e+04
open_porch_sf             9.646022e+03
ms_subclass_40            7.780165e+03
ms_subclass_Others        7.026999e+03
total_bsmt_sf             6.777101e+03
ms_zoning_Others          6.186262e+03
overall_qual              5.712288e+03
ms_subclass_180           4.340172e+03
ms_subclass_50            4.063148e+03
ms_subclass_70            3.964275e+03
exterior_1st_MetalSd      3.786049e+03
exterior_2nd_HdBoard      3.750420e+03
bsmt_qual                 3.577336e+03
gr_liv_area               3.259363e+03
year_remod_add            3.189192e+03
year_built                3.012803e+03
ms_subclass_160           2.141524e+03
totrms_abvgrd             1.746011e+03
ms_subclass_80            1.223862e+03
wood_deck_sf              1.069790e+03
ms_subclass_90            8.639973e+02
ms_zoning_RL              8.143879e+02
ms_subclass_190           4.878902e+02
ms_subclass_45            2.709896e+02
roof_style_Gable          2.428104e+02
kitchen_qual              1.435052e+02
ms_subclass_120           3.655727e-11
exterior_2nd_Others       1.401779e-12
exterior_2nd_VinylSd      8.381250e-14
ms_subclass_75           -0.000000e+00
lot_frontage_reg_imputed  0.000000e+00
mas_vnr_type_None         0.000000e+00
mas_vnr_type_CBlock      -0.000000e+00
lot_shape                 0.000000e+00
1st_flr_sf                0.000000e+00
lot_config_Others         0.000000e+00
full_bath                -0.000000e+00
bsmtfin_sf_1              0.000000e+00
exterior_1st_HdBoard     -0.000000e+00
exterior_1st_Others       0.000000e+00
exterior_1st_PreCast     -0.000000e+00
exterior_1st_VinylSd      0.000000e+00
exterior_1st_Wd Sdng     -0.000000e+00
exterior_2nd_Wd Sdng      0.000000e+00
fence                    -0.000000e+00
fireplace_qu              0.000000e+00
bsmt_exposure             0.000000e+00
foundation_Others         0.000000e+00
fireplaces               -0.000000e+00
lot_config_Inside         0.000000e+00
heating_qc                0.000000e+00
garage_cars               0.000000e+00
garage_finish             0.000000e+00
garage_type_Attchd        0.000000e+00
garage_type_Others        0.000000e+00
lot_area                  0.000000e+00
house_style_Others       -0.000000e+00
house_style_1Story        0.000000e+00
ms_subclass_30           -8.840224e+01
neigh_group              -3.180020e+02
bsmtfin_type_1           -5.332833e+02
exter_qual               -7.891973e+02
exterior_2nd_MetalSd     -1.082117e+03
ms_subclass_60           -2.123081e+03
exterior_2nd_Other       -4.072848e+03
exterior_2nd_PreCast     -4.651278e+03
mas_vnr_type_Others      -5.448638e+03
mas_vnr_area             -6.328407e+03
garage_area              -9.401700e+03

In [28]:
#Number of coefficient remaining after fitting
lasso_coef[lasso_coef!=0].notna().sum()

Coefficient    43
dtype: int64

We see that the lasso regularization has reduced the previously overblown coefficients. The number of coefficients also reduced more significantly from 75 to 42 through shutting off the coefficients.

### Perform Elastic Net

Lastly, we will train and evaluate two models of ElasticNet

#### Elastic Net with l1_ratio  manually selected at 0.2

In [29]:
enet_ratio = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]

enet_l120 = ElasticNetCV(l1_ratio=0.2, n_alphas=1000)

enet_l120.fit(X_train_ss, y_train)

print(round(enet_l120.alpha_, 2))
print(enet_l120.l1_ratio_)

322.57
0.2


In [30]:
round((-cross_val_score(
    enet_l120,
    X_train_ss,
    y_train,
    cv=10,
    scoring='neg_mean_squared_error'
).mean())**0.5, 2)

77045.76

In [31]:
round((mean_squared_error(y_test, enet_l120.predict(X_test_ss)))**0.5, 2)

73818.45

In [32]:
enet_l120_coef = pd.DataFrame(enet_l120.coef_, 
                          index = X.columns, 
                          columns=['Coefficient']
                         ).sort_values(by='Coefficient', ascending=False)

enet_l120_coef

Coefficient
roof_style_Others          240.621067
ms_subclass_40             214.047119
ms_subclass_85             213.348850
ms_subclass_Others         205.465081
total_bsmt_sf              201.863347
overall_qual               197.066895
ms_subclass_70             195.641275
ms_subclass_160            194.360187
ms_subclass_75             191.784121
ms_subclass_190            182.678557
year_built                 169.906482
ms_subclass_80             166.974338
ms_subclass_50             162.011740
year_remod_add             161.843177
ms_subclass_60             158.215628
open_porch_sf              155.868629
totrms_abvgrd              152.190690
ms_subclass_20             144.010672
ms_subclass_90             134.261724
ms_subclass_180            128.912759
ms_subclass_30             112.306895
ms_zoning_Others           107.694768
ms_zoning_RL               103.849469
wood_deck_sf               101.723265
roof_style_Gable            97.332468
exterior_2nd_MetalSd        92.443178
exterior_2nd_Others         78.561193
full_bath                   58.098156
exterior_2nd_PreCast        53.632854
bsmtfin_type_1              50.664089
exterior_1st_Wd Sdng        50.442547
kitchen_qual                42.255808
ms_subclass_120             29.840233
mas_vnr_type_None           28.099621
fireplaces                  11.570700
garage_type_Others          10.402258
garage_area                  6.946663
fence                        4.066729
lot_shape                    0.000000
lot_config_Inside            0.000000
foundation_Others            0.000000
heating_qc                   0.000000
exterior_1st_VinylSd         0.000000
garage_cars                  0.000000
garage_finish                0.000000
garage_type_Attchd           0.000000
exterior_1st_Others          0.000000
house_style_1Story           0.000000
lot_area                     0.000000
bsmtfin_sf_1                 0.000000
lot_config_Others            0.000000
lot_frontage_reg_imputed     0.000000
bsmt_qual                   -0.518142
exterior_2nd_Wd Sdng        -4.066729
exterior_1st_PreCast        -5.448127
exterior_1st_HdBoard       -11.095948
fireplace_qu               -11.570700
1st_flr_sf                 -12.047829
house_style_Others         -16.007142
gr_liv_area                -16.094754
exterior_1st_MetalSd       -16.163686
exterior_2nd_HdBoard       -16.984785
bsmt_exposure              -17.432607
exter_qual                 -19.915510
mas_vnr_area               -27.502834
mas_vnr_type_CBlock        -28.099620
mas_vnr_type_Others        -29.840233
ms_subclass_45             -49.761304
exterior_2nd_VinylSd       -53.632855
foundation_PConc           -58.098155
exterior_2nd_Other         -78.561193
neigh_group                -95.309326

In [33]:
#Number of coefficient remaining after fitting
enet_l120_coef[enet_l120_coef!=0].notna().sum()

Coefficient    58
dtype: int64

The ElasticNet model with l1_ratio at 0.2 exhibited more signifcant regularization of coefficients, but this comes at the expense of poorer performance in terms of RMSE. The number The number of coefficients reduced from 75 to 58 after fitting.

#### Elastic Net with l1_ratio decided by ElasticNetCV

In [34]:
enet_ratio = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]

enet_model = ElasticNetCV(l1_ratio=enet_ratio, n_alphas=1000)

enet_model.fit(X_train_ss, y_train)

print(round(enet_model.alpha_, 2))
print(enet_model.l1_ratio_)

71.68
0.9


In [35]:
round((-cross_val_score(
    enet_model,
    X_train_ss,
    y_train,
    cv=10,
    scoring='neg_mean_squared_error'
).mean())**0.5, 2)

43088.42

In [36]:
round((mean_squared_error(y_test, enet_model.predict(X_test_ss)))**0.5, 2)

42049.49

In [37]:
enet_coef = pd.DataFrame(enet_model.coef_, 
                          index = X.columns, 
                          columns=['Coefficient']
                         ).sort_values(by='Coefficient', ascending=False)

enet_coef

Coefficient
roof_style_Others         3972.095799
ms_subclass_85            3732.780016
ms_subclass_40            3387.799795
ms_subclass_Others        3273.620254
total_bsmt_sf             3260.334607
ms_subclass_160           3194.416891
ms_subclass_70            2930.473459
overall_qual              2898.443707
ms_subclass_75            2726.186114
open_porch_sf             2660.234397
ms_subclass_190           2652.197600
totrms_abvgrd             2613.675390
ms_subclass_50            2596.786211
ms_subclass_20            2485.459683
ms_subclass_80            2255.243726
year_remod_add            2227.192967
ms_subclass_60            2225.847967
year_built                2166.766717
ms_subclass_180           1990.873573
ms_zoning_Others          1892.763701
ms_subclass_90            1883.805452
ms_subclass_30            1649.301657
ms_zoning_RL              1627.948935
wood_deck_sf              1507.333198
exterior_2nd_MetalSd      1492.873886
roof_style_Gable          1469.177903
exterior_2nd_Others       1097.547509
full_bath                  753.631553
exterior_2nd_PreCast       643.797727
kitchen_qual               636.407152
exterior_1st_Wd Sdng       613.263002
bsmtfin_type_1             605.994738
ms_subclass_120            578.912218
mas_vnr_type_None          355.082383
fence                      157.305158
fireplaces                 156.438806
garage_type_Others         104.784148
bsmt_qual                   58.152513
garage_area                 35.085558
lot_config_Inside            0.000000
exterior_1st_VinylSd         0.000000
house_style_1Story           0.000000
foundation_Others            0.000000
exterior_1st_Others          0.000000
garage_cars                  0.000000
garage_finish                0.000000
garage_type_Attchd           0.000000
heating_qc                   0.000000
lot_area                     0.000000
lot_config_Others            0.000000
lot_frontage_reg_imputed     0.000000
bsmtfin_sf_1                 0.000000
lot_shape                    0.000000
exterior_1st_PreCast       -33.753333
house_style_Others        -140.488197
exterior_1st_MetalSd      -152.445005
fireplace_qu              -156.438716
exterior_2nd_Wd Sdng      -157.305761
exterior_2nd_HdBoard      -171.083402
bsmt_exposure             -175.289495
gr_liv_area               -210.540716
exterior_1st_HdBoard      -228.980869
exter_qual                -236.276385
1st_flr_sf                -243.580846
mas_vnr_type_CBlock       -355.082311
mas_vnr_area              -425.248156
mas_vnr_type_Others       -578.911976
exterior_2nd_VinylSd      -643.798444
ms_subclass_45            -664.162890
foundation_PConc          -753.631033
exterior_2nd_Other       -1097.546835
neigh_group              -1346.661753

In [38]:
#Number of coefficient remaining after fitting
enet_coef[enet_coef!=0].notna().sum()

Coefficient    58
dtype: int64

The ElasticNet model hyperparameters chosen by ElasticNetCV exhibited less regularization of coefficients compared to the one where l1_ration is manually set, but performs better in terms of RMSE. However, it still performs worse than the Linear, Lasso and Ridge Regression. The number of coefficients reduced from 75 to 58 after fitting.

### Model Selection

The following summarizes the models trained and their performance:

| Models | Description       | Hyperparams               | Features | CV RMSE  | Holdout RMSE |
|--------|-------------------|---------------------------|----------|----------|--------------|
| 1      | Elastic Net       | alpha 322.57 l1 ratio 0.2 | 58       | 77045.76 | 73818.45     |
| 2      | Elastic Net       | alpha 71.68 l1 ratio 0.9  | 58       | 43088.42 | 42049.49     |
| 3      | Lasso Regression  | alpha 92.43               | 42       | 25831.85 | 28636.59     |
| 4      | Ridge Regression  | alpha 29.15               | 58       | 29336.05 | 28669.68     |
| 5      | Linear Regression | -                         | 71       | 25928.04 | 28717.78     |

The best performing model is the Lasso Regression with alpha at 92.43. We will proceed to instantiate LassoRegression at alpha = 92.43 and re-train the entire training dataset and use it to predict the Kaggle test set for submission to Kaggle.

In [39]:
# Create X_test from df_test
X_test = df_test.drop(columns=['id', 'pid'])

In [40]:
# Isolate continuous variables for scaling on both the full train dataset X and X_tes
X_cont = X[con_var]
X_test_cont = X_test[con_var]

In [41]:
#Scale the continuous variables
X_ss_cont = ss.fit_transform(X_cont)
X_test_ss_cont = ss.transform(X_test_cont)

In [42]:
# Drop the orginal continuous variables from X and join back the scaled ones
X_ss = X.drop(
    columns=con_var
    ).reset_index(drop=True
                 ).join(pd.DataFrame(X_ss_cont, columns=con_var))

In [43]:
X_ss

exterior_1st_HdBoard  exterior_1st_MetalSd  exterior_1st_Others  \
0                      1.0                   0.0                  0.0   
1                      0.0                   0.0                  0.0   
2                      0.0                   0.0                  0.0   
3                      0.0                   0.0                  0.0   
4                      0.0                   0.0                  0.0   
...                    ...                   ...                  ...   
2035                   0.0                   0.0                  0.0   
2036                   0.0                   0.0                  0.0   
2037                   0.0                   0.0                  1.0   
2038                   0.0                   0.0                  1.0   
2039                   0.0                   0.0                  0.0   

      exterior_1st_PreCast  exterior_1st_VinylSd  exterior_1st_Wd Sdng  \
0                        0                   0.0                   0.0   
1                        0                   1.0                   0.0   
2                        0                   1.0                   0.0   
3                        0                   1.0                   0.0   
4                        0                   0.0                   1.0   
...                    ...                   ...                   ...   
2035                     0                   1.0                   0.0   
2036                     0                   1.0                   0.0   
2037                     0                   0.0                   0.0   
2038                     0                   0.0                   0.0   
2039                     0                   1.0                   0.0   

      exterior_2nd_HdBoard  exterior_2nd_MetalSd  exterior_2nd_Other  \
0                      0.0                   0.0                   0   
1                      0.0                   0.0                   0   
2                      0.0                   0.0                   0   
3                      0.0                   0.0                   0   
4                      0.0                   0.0                   0   
...                    ...                   ...                 ...   
2035                   0.0                   0.0                   0   
2036                   0.0                   0.0                   0   
2037                   0.0                   0.0                   0   
2038                   0.0                   0.0                   0   
2039                   0.0                   0.0                   0   

      exterior_2nd_Others  exterior_2nd_PreCast  exterior_2nd_VinylSd  \
0                     1.0                     0                   0.0   
1                     0.0                     0                   1.0   
2                     0.0                     0                   1.0   
3                     0.0                     0                   1.0   
4                     1.0                     0                   0.0   
...                   ...                   ...                   ...   
2035                  0.0                     0                   1.0   
2036                  0.0                     0                   1.0   
2037                  1.0                     0                   0.0   
2038                  1.0                     0                   0.0   
2039                  0.0                     0                   1.0   

      exterior_2nd_Wd Sdng  fireplaces  foundation_Others  foundation_PConc  \
0                      0.0           0                1.0               0.0   
1                      0.0           1                0.0               1.0   
2                      0.0           0                1.0               0.0   
3                      0.0           0                0.0               1.0   
4                      0.0           0                0.0               1.0   
...                    .

In [44]:
# Drop the orginal continuous variables from X_test and join back the scaled ones
X_test_ss = X_test.drop(
    columns=con_var
    ).reset_index(drop=True
                 ).join(pd.DataFrame(X_test_ss_cont, columns=con_var))

In [45]:
# Instantiate Lasso with alpha = 92.43 and fit on X_ss and y
lasso_final = Lasso(alpha = 92.43)

lasso_final.fit(X_ss,y)

Lasso(alpha=92.43)

In [46]:
# Predict on the Kaggle test set
y_pred = lasso_final.predict(X_test_ss)

#### Coefficient of Continuous Variables

In [47]:
lasso_final_coef = pd.DataFrame(lasso_final.coef_, index = X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)
coef_cont_var = lasso_final_coef.loc[lasso_final_coef.index.isin(con_var), :]
coef_cont_var.loc[coef_cont_var['Coefficient']!=0, :]

Coefficient
open_porch_sf   8.229412e+03
total_bsmt_sf   5.316694e+03
overall_qual    5.093659e+03
bsmt_qual       3.827485e+03
gr_liv_area     3.460017e+03
year_remod_add  3.278613e+03
year_built      1.797624e+03
totrms_abvgrd   1.007716e+03
wood_deck_sf    8.807518e+02
fireplace_qu    5.220919e+02
full_bath      -4.546635e-11
exter_qual     -4.844431e+03
mas_vnr_area   -6.507661e+03
garage_area    -6.959350e+03

#### Coefficient of Categorical Variables

In [48]:
coef_cat_var = lasso_final_coef.loc[~lasso_final_coef.index.isin(con_var), :]
coef_cat_var.loc[coef_cat_var['Coefficient']!=0, :]

Coefficient
ms_subclass_85        2.177132e+04
roof_style_Others     1.705238e+04
ms_subclass_20        1.106050e+04
foundation_PConc      9.322009e+03
ms_subclass_40        8.342410e+03
ms_subclass_Others    6.839128e+03
ms_zoning_Others      5.547894e+03
ms_subclass_180       5.217167e+03
ms_subclass_70        4.967097e+03
ms_subclass_50        3.480368e+03
exterior_2nd_HdBoard  3.251143e+03
garage_type_Others    2.336174e+03
ms_subclass_160       2.127763e+03
ms_subclass_90        1.326125e+03
ms_subclass_80        1.046495e+03
roof_style_Gable      9.378328e+02
ms_subclass_190       6.270954e+02
exterior_1st_MetalSd  5.720949e+02
ms_subclass_45        3.131904e+02
exterior_2nd_VinylSd  3.418691e-12
ms_subclass_120       7.758193e-13
fireplaces           -1.020166e-12
mas_vnr_type_CBlock  -3.116590e+02
exterior_1st_PreCast -9.165129e+02
ms_subclass_30       -1.299562e+03
exterior_1st_HdBoard -1.683132e+03
ms_subclass_60       -1.925286e+03
exterior_2nd_Other   -3.407769e+03
mas_vnr_type_Others  -5.410518e+03
exterior_2nd_PreCast -5.422062e+03

In [49]:
# Preparation for submission to Kaggle
df_test['saleprice'] = y_pred

In [50]:
submit = df_test[[col for col in df_test.columns if col in['id', 'saleprice']]]

In [51]:
submit.columns = ['Id', 'SalePrice']

In [52]:
submit.to_csv('../data/kaggle_submit.csv', index=False)

### Conclusion and Recommendation

The submitted Lasso Regression with alpha = 92.43 model has a RMSE score of about \\$30,430 on Kaggle. The holdout RMSE is \\$28,636 for this model in comparison. This is in line with expectation, where the model do slightly worse on data it has not seen.

We may infer the influence of each features from the coefficient of this model. It is necessary to discuss the influence of features according to whether they are categorical or continuous in nature.

**Categorical Features**
1. The model show that houses which are of the categories MS SubClass 20, 40, 85 and 180 can command a premium of \\$5000 to \\$21000. These four classes are 1-STORY 1946 & NEWER ALL STYLES, 1-STORY W/FINISHED ATTIC ALL AGES, SPLIT FOYER and PUD - MULTILEVEL - INCL SPLIT LEV/FOYER respectively.

2. Houses with foundation made of poured concrete also commands higher prices of at least \\$9000

3. The presence of precast coverings lowers the sale price of a house by at least $5000

**Continuous Features**

1. Big open porches, total basement and ground living areas have a positive influence on house price, with price increase ranging from \\$3,000 to \\$9,000 per square foot.

2. Quite intuitively, quality related features such as overall finishing quality and height of basement are positive predictor of house price

3. A huge garage area negatively impacts the value of a house, decreasing by almost \\$7000 dollars for every square foot.

4. As expected, house price decreases with age and year since remod. 

**Recommendation and Further Study**

1. Homeowners in Ames should consider expanding their own open porch if they do not already have one to enhance their property value. The cost of building a porch is a 200-square-foot is between \\$4600 to \\$22000. [link](https://www.homeadvisor.com/cost/outdoor-living/build-a-porch/). The benefits clearly outweighs the cost.

2. Homeseekers would be advised to avoid houses with a huge a garage and and has precast coverings due to these features drag on the property value.

3. While location is a known predictor of property price, the Lasso Regression model has penalised the coefficient of the `neigh_group` feature (whereby neighborhood were grouped into 5 groups according to house sale price) to zero, therefore eliminating its influence. More specific information on the exact addresses of the property instead of neighborhoods would be helpful to make the Lasso Regression model more predictive.

4. Most of the findings in this project are related to the features of the property itself, but it is known that economic conditions and interest rate are factors influencing property prices. The boxplot of `yr_sold` vs `saleprice` did not surface any significant change in distribution of the `saleprice` from 2006 to 2010 despite 2009 to 2010 being the years of the Global Financial Crisis. Perhaps, more data on the prevailing interest rates can be retrieved to see its correlation with housing price.